libcxx/include/__ranges/concat_view.h

```diff
@@ -0,0 +1,638 @@
+                      __iy, [&](auto __index_constant_y) {
+                        constexpr size_t __index_y = __index_constant_y.value;
+                        auto __dy =
+                            ranges::__distance(ranges::begin(std::get<__index_y>(__y.__parent_->__views_)), __it_y);
+                        difference_type __s = 0;
+                        for (size_t __idx = __index_y + 1; __idx < __index_x; __idx++) {
+                          __s += ranges::size(std::get<__idx>(__x.__parent_->__views_));
+                        }
+                        return __dy + __s + __dx;
+                      });
+                });
+          },
+          __x.__it_,
+          __y.__it_);
+    } else if (__ix < __iy) {
+      return -(__y - __x);
+    } else {
+      std::visit([&](const auto& __it1, const auto& __it2) { return __it1 - __it2; }, __x.__it_, __y.__it_);
+    }
+  }
+
+  _LIBCPP_HIDE_FROM_ABI friend constexpr __iterator operator-(const __iterator& __it, difference_type __n)
+    requires __concat_is_random_access<_Const, _Views...>
+  {
+    auto __temp = __it;
+    __temp -= __n;
+    return __temp;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI friend constexpr difference_type operator-(const __iterator& __x, default_sentinel_t)
+    requires(sized_sentinel_for<sentinel_t<__maybe_const<_Const, _Views>>, iterator_t<__maybe_const<_Const, _Views>>> &&
+             ...) &&
+            (__apply_drop_first<_Const, _Views...>::value)
+  {
+    size_t __ix = __x.__it_.index();
+    std::visit(
+        [&](auto& __it_x) {
+          __it_x.template __apply_at_index<tuple_size_v<decltype(__x.__parent_->__views_)>>(
+              __ix, [&](auto __index_constant) {
+                constexpr size_t __index_x = __index_constant.value;
+                auto __dx = ranges::__distance(ranges::begin(std::get<__index_x>(__x.__parent_->__views_)), __it_x);
+
+                difference_type __s = 0;
+                for (size_t __idx = 0; __idx < __index_x; __idx++) {
+                  __s += ranges::size(std::get<__idx>(__x.__parent_->__views_));
+                }
+
+                return -(__dx + __s);
+              });
+        },
+        __x.__it_);
+  }
+
+  _LIBCPP_HIDE_FROM_ABI friend constexpr difference_type operator-(default_sentinel_t, const __iterator& __x)
+    requires(sized_sentinel_for<sentinel_t<__maybe_const<_Const, _Views>>, iterator_t<__maybe_const<_Const, _Views>>> &&
+             ...) &&
+            (__apply_drop_first<_Const, _Views...>::value)
+  {
+    -(__x - default_sentinel);
+  }
+};
+
+namespace views {
+namespace __concat {
+struct __fn {
+  template <class... _Views>
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr auto operator()(_Views... __views) const
+      noexcept(noexcept(concat_view(std::forward<_Views>(__views)...)))
+          -> decltype(concat_view(std::forward<_Views>(__views)...)) {
+    return concat_view(std::forward<_Views>(__views)...);
+  }
+};
```

Comment:
It seems that the `views::all` case in [[range.concat.overview]](https://eel.is/c++draft/range.concat.overview) was completely ignored.

IMO it would be helpful to write a helper `constexpr` variable template to record both strategy (__all, __concat, or __none) and noexcept status.